### ISSUE TITLE
x/tools/packages: Load does not respect Go version consistently

### ISSUE THREAD

**Issue by hugelgupf** [golang/go#62114]
x/tools/packages invokes the "go" command, but often behaves differently depending on the underlying Go version. As a consumer of the package, I regularly test my usage against several versions of Go, because these tools behave differently depending on the underlying Go version.

At one point, I believed that setting `packages.Config.Env` with the correct GOROOT and PATH=$GOROOT/bin:$PATH would make up for this, but because `exec.Command` uses `exec.LookPath` (which is hard-coded to the system `os.Getenv("PATH")`, see #20081, #59753, #60601) and `packages.Load` does nothing to mitigate that, the Go version used to load packages is not the correct one.

Here's an example usage of packages.Load:

```go
package main

import (
        "fmt"
        "go/build"
        "log"
        "os"
        "runtime"

        "golang.org/x/tools/go/packages"
)

func GOROOT() string {
        if g := build.Default.GOROOT; g != "" {
                return g
        }
        return runtime.GOROOT()
}

func main() {
        fmt.Println("GOROOT:", GOROOT())
        fmt.Println("PATH:", os.Getenv("PATH"))
        cfg := &packages.Config{
                Mode: packages.LoadSyntax | packages.LoadImports,
                Env: []string{
                        "GOPACKAGESDRIVER=off",
                        // Packages driver needs a GOROOT to look up cmd/test2json.
                        fmt.Sprintf("GOROOT=%s", GOROOT()),
                        // Where to find the Go binary itself.
                        fmt.Sprintf("PATH=%s/bin", GOROOT()),
                        // Package driver needs HOME (or GOCACHE) for build cache.
                        fmt.Sprintf("HOME=%s", os.Getenv("HOME")),
                },
                Dir: "",
        }
        pkgs, err := packages.Load(cfg, "cmd/test2json")
        if err != nil {
                log.Fatal(err)
        }
        for _, pkg := range pkgs {
                fmt.Printf("%#v\n", pkg)
        }
}
```

Say you are on a system where the admin has installed some older Go, but you've used `golang.org/dl/go1.20` to download a newer version. Or it's a newer version of Go, but you want to ensure that your `packages.Load` code works against every Go version 1.18 and up.

```
$ go1.20 build -o pkgdriver
$ GOROOT=$HOME/sdk/go1.20 strace -fe trace=execve -e signal=SIGIO ./pkgdriver 2>&1 | grep -v "exited with" | grep -v "strace: Process"

[pid 675271] execve("/usr/lib/google-golang/bin/go", ["go", "list", "-e", "-f", "{{context.ReleaseTags}}", "--", "unsafe"], 0xc00007e340 /* 6 vars */) = 0                 
[pid 675272] execve("/usr/lib/google-golang/bin/go", ["go", "list", "-f", "{{context.GOARCH}} {{context.Com"..., "--", "unsafe"], 0xc00009c150 /* 5 vars */) = 0           
[pid 675282] execve("/usr/lib/google-golang/bin/go", ["go", "list", "-e", "-json=Name,ImportPath,Error,Dir,"..., "-compiled=true", "-test=false", "-export=true", "-deps=tr
ue", "-find=false", "-pgo=off", "--", "cmd/test2json"], 0xc00002ea50 /* 5 vars */) = 0                                                                                     
[pid 675291] execve("/usr/local/google/home/chrisko/sdk/go1.20/pkg/tool/linux_amd64/compile", ["/usr/local/google/home/chrisko/s"..., "-V=full"], 0xc0001d9d40 /* 25 vars *
/) = 0                                                                                                                                                                     
[pid 675292] execve("/usr/local/google/home/chrisko/sdk/go1.20/pkg/tool/linux_amd64/compile", ["/usr/local/google/home/chrisko/s"..., "-V=full"], 0xc00013cc30 /* 25 vars *
/ <unfinished ...>
...
```

Despite best efforts, the system "go" (which might be older) is used to execute the list driver, but $GOROOT/bin/go is then used to compile more information about the package by calling compile/asm functions. This is inconsistent.

Note that if instead of compiling the command and then running it, you run it with `go1.20 run`, the command seems to mitigate this itself by placing GOROOT/bin in the PATH prior to execution:

```
$ go1.20 run .
PATH: /usr/local/google/home/chrisko/sdk/go1.20/bin:<snip>
```

**Comment by dmitshur** [comment]
CC @matloob, @golang/tools-team.

**Comment by hugelgupf** [comment]
> Is the request that we essentially look up the path to the go command using the given environment before running it?

Yes. Especially because depending on the situation, the Go packages driver code either directly calls the go command (in which case, system-PATH-go is used) or first invokes a gopackagesdriver that may the same thing (but which now has the PATH set to what I want, and therefore uses the go binary that I want). And if the system-PATH-go command invokes go again, it will _then_ use the go supplied in the PATH I set (This is what you can see in the strace example above -- `google/go list` invoked by API fork/execs `go1.20/go compile`, because the PATH changed).

The outcome of a Load call is different depending on the nesting of subprocesses that is invoked by the implementation of Load, which IMO leads to an API that returns confusing results.

> Would you be able to set the environment on the tool that you run so that the correct go command is in the path when you run the tool?

I'm supplying these tools as both binaries and APIs. It's possible in the binaries, it's not really possible in the APIs, as I'd have to mess with the caller's PATH for the entire process.

**Comment by hugelgupf** [comment]
I think another option, rather than using the PATH supplied in env, would be an argument that gives the ability to specify the `go` command location.

**Comment by matloob** [comment]
I want to push back a little on this. Right now the contract is that the user of the binary that uses go/packages ensures that the right go command is in their PATH before running it. I think it's reasonable to require users of your API to make sure that they have the correct version of go in their PATH.

What's the use case for using a different version of Go when using go/packages through an API you provide?

**Comment by dmitshur** [comment]
> the binary name that is in PATH is go1.21, not go.

Those goN.M commands modify PATH in the environment to put their own "go" binary first in the list. See [here](https://cs.opensource.google/go/dl/+/master:internal/version/version.go;l=63-66;drc=2a66dfb9a9cf29fab51425f5c052d40437875c70).

**Comment by hugelgupf** [comment]
Indeed, but you only get that if your tool using go/packages is run with
goN.M run. If you build it with goN.M and then run the binary, you have to
supply a PATH and a GOROOT to that binary to get the correct behavior.

**Comment by hyangah** [comment]
How about using the new `GOTOOLCHAIN` environment variable?
One caveat is that the version should be the exact go toolchain version, not the language version.

```
$ go version
go version go1.21.5 darwin/amd64
$ GOTOOLCHAIN=go1.19.5 go version
go version go1.19.5 darwin/amd64
```

**Comment by StevenACoffman** [comment]
This problem seems to be exacerbated when you use `go run` to run a Go CLI tool and there are discrepancies in the local Go version, toolchain version, and `go.mod` versions (for both the CLI tool and the local code base it is operating on).

For instance, I maintain [gqlgen](https://github.com/99designs/gqlgen), a GraphQL server codegen CLI tool which is also a library for the resulting generated code.

People have executed gqlgen with `go run -mod=mod github.com/99designs/gqlgen@v0.17.62` (which requires Go 1.22.5), and it will call `packages.Load` to look for local Go package files. Users then report that `packages.Load` for a list of local Go files are not populated (printing it results in `&{map[] map[] map[] map[] map[] map[] map[] [] map[]}`).

Changing the packages.Load config to have `&packages.Config{..., Env: append(os.Environ(), "GOTOOLCHAIN=go1.23.4")` does not resolve the problem, but it works fine if instead the GOTOOLCHAIN environment variable is set prior to execution:
```
export GOTOOLCHAIN=1.23.4
go run -mod=mod github.com/99designs/gqlgen@v0.17.62 --config gqlgen.yml
```
I have not tried `err := os.Setenv("GOTOOLCHAIN","go1.23.4")` prior to `packages.Load` as it seems tricky to determine which Go version I should set it to.

I would be grateful for any advice.

**Comment by dmitshur** [comment]
@StevenACoffman What you’re describing may not necessarily be the same issue. This issue seems to be about inconsistency in go version used when “go1.N.M” binaries are used to run vs to build Go commands, because those commands temporarily modify the environment. What you’re describing sounds like a case of packages.Load not loading packages rather than an inconsistent version being used.

It might be that packages.Load is working as intended under that situation (e.g., perhaps the packages being loaded are not present in the build graph under that invocation), or there may be a problem. It should help if you file a separate issue with enough information to be able to reproduce the behavior you’re reporting. Thanks.

**Comment by StevenACoffman** [comment]
Thanks! I'll try to make a simpler reproduction case (with out all the layers) to see if that makes it more obvious what is going on.

**Comment by mwriter** [comment]
@hugelgupf 

Oh, you ran into this problem much earlier than me. And there it comes to generally funny things:
```
p := "some_path"
os.Setenv("PATH", p)
cmd := exec.Command("some_binary_in_some_path")
err := cmd.Start() // <nil>
```
```
p := "some_path"
cmd := exec.Command("some_binary_in_some_path")
os.Setenv("PATH", p)
err := cmd.Start() // exec: "some_binary_in_some_path": executable file not found in $PATH
```
As a result, I completely rewrote `exec.Cmd` for myself.
But in this particular case, you can go for such a trick:
```
PATH="path_to_toolchain_1:...:path_to_toolchain_I:$PATH" GOTOOLCHAIN="goM.N+path" go run .
```
Or try to use `GOTOOLCHAIN` in `packages.Config.Env` and
```
go build . -o out
PATH="path_to_toolchain_1:...:path_to_toolchain_I:$PATH" out
```
Although in my opinion, there are still enough other problems in the package `x/tools/go/packages`.

**Comment by oakad** [comment]
A great diversity of "duplicate" issues now point here. Unfortunately, this particular issue is not obviously named to reflect the real underlying problem with go command invocation by "x/tools". May it be worthwhile to rename this issue or to actually open another, active, "not dup" issue called "make x/tools invoke go binary in a properly configurable fashion consistent with modern security and use cases"? :-)

**Comment by adonovan** [comment]
Although this is nominally about packages.Load, it's really about the behavior of os/exec.Command("go") and PATH, so I'm going to mark this as a Go command issue and loop in its working group. I'm happy to make whatever changes are needed to packages.Load but I'm not sure what they are.

**Comment by mwriter** [comment]
@adonovan 

This is a question that probably shouldn't be asked in my position, but I'll ask it anyway, especially since you yourself advised me to discuss everything here.
Why is my comment worse than the last comment from `oakad`? And if nothing, why is it hidden?
I have quite a few comments that have the same meaning and connotation as the comment from Alex, but my comments are hidden.

**Comment by adonovan** [comment]
@mwriter I didn't see your comment till now, and I don't know who hid it or what their specific reasons were. It is a little hard to follow, but it doesn't seem egregiously low quality.
 
- It points out a known subtlety of exec.Command, that it resolves relative executable paths eagerly during construction, which is useful and relevant.
- Suggesting (even implicitly) rewriting exec.Command is obviously not really a solution nor something to encourage.
- The "trick" is rather confusing.
- The final barb about go/packages is neither constructive nor informative.

If I had to guess, I suspect it was the final remark that got the post flagged. But go/packages does have several problems, and we can have a thicker skin than that. I'll un-hide it.

**Comment by mwriter** [comment]
@adonovan 

It seems to me that the problem lies elsewhere, and I will try to voice it. And I hope the team will also be understanding.
I once again publicly apologize to `Sean` for my attacks on him last year. But I also want to help make the language better here.
If he has specific comments on my messages, we can always discuss it in person.
But please don't use the presumption of guilt on me and exclude me from the discussions.
PS. I'll hide this message myself soon.

**Comment by seankhliao** [comment]
I'm not sure what else cmd/go is expected to do? it typically arranges for the path element that it comes from to be first, in the case of go run, tool, generate, test.

In the case of go/packages when invoked by some external process, it seems it would be up to that process to arrange for the right go to be called.

**Comment by seankhliao** [comment]
see for example #68005 

I think if using a wrapper like dl/go1.x.y, you need to skip it and add the underlying GOROOT/bin path into your path, so `go` directly works at the version you want. 

Or use GOTOOLCHAIN instead.

In other words, https://github.com/golang/go/issues/62114#issuecomment-1881730220 there is nothing to do here, besides better documentation on there expected environment.

**Comment by mwriter** [comment]
> it typically arranges for the path element that it comes from to be first

But why is the path not derived from the specified environment in `packages.Config`?
But, again, I think this is an issue of `exec.Cmd`.

**Comment by oakad** [comment]
@seankhliao With Bazel.Build (and similar systems), there's a general desire to reduce the number of unknowns when invoking binaries. Toolchain, in many cases, will be local to the build; the build itself may be happening on some remote node.
Yes, everything can be resolved with wrapper scripts and launch helpers - that's how we use those things right now. But,
1. Launch helpers pollute already complicated projects.
2. Presence of configurable options in the "quick help"  signals the users that some command may need to invoke additional executables. For example, my first few attempts to invoke goimports from Bazel rule were quite frustrating, because it was not obvious at all it will try to invoke "go" and fail in non-obvious fashion being unable to do so.

**Comment by dmitshur** [comment]
From triage, we discussed this and believe the go command itself is working as intended. It's expected that the intended version of the go command needs to be first in PATH environment variable. We're leaving this as an x/tools/go/packages issue, though there may not be much that can be done there either.

**Comment by dmitshur** [comment]
> I'm happy to make whatever changes are needed to packages.Load but I'm not sure what they are.

@adonovan The documentation at https://pkg.go.dev/golang.org/x/tools/go/packages#Config.Env includes "Env is the environment to use when invoking the build system's query tool.", but it doesn't discuss how the build system's query tool is found. There is a brief mention at https://pkg.go.dev/golang.org/x/tools/go/packages#DriverResponse, "the driver (e.g. the go command on the PATH)".

Given it can be a misuse of the API to try and call Packages.Load with Config.Env that contains a modified PATH in an attempt to get packages.Load to select another version of the go command, maybe it can be detected and reported as an error? Something like:

```shell
# (in a terminal)
$ go build -o /tmp/a
$ export PATH="/specific/go1.n.m/bin:$PATH"  # provide a specific version of the go binary to the /tmp/a program
$ /tmp/a
```

```go
// (the "go" binary dependency is already in PATH of the process environment)
{
	packages.Load(cfg, &packages.Config{Env: nil}) // okay, finds intended version of "go" binary in PATH
	anotherapi.Load(...) // if this API also has the "go" binary in PATH as a dependency, it'll select the same version as packages.Load did
}

{
	_, err := packages.Load(cfg, &packages.Config{Env: []string{"PATH=/another/go1.n.m/bin"})
	// err is intended to be a helpful error pointing out that "/another/go1.n.m/bin" contains a different version of the go binary than the one in the parent process
}
```

If it's not viable to report an error (e.g., due to a risk of false positives), maybe it would help to expand Config.Env documentation to mention that it's an ineffective way to modify which "go" command is first in PATH.

**Comment by mwriter** [comment]
> It's expected that the intended version of the go command needs to be first in PATH environment variable.

The question is, in what kind of environment. I still categorically don't understand why the commands accept `PATH` only from the global environment.

```
$ which whoamibin
```
```
$ which whoami
/bin/whoami
```
```
$ PATH=/system/bin which whoami
/system/bin/whoami
```
```
$ PATH= which whoami
sh: which: No such file or directory
```
But for some reason, in the concept of `exec.Cmd`:
`$ which whoami` == `$ PATH=/system/bin which whoami` == `$ PATH= which whoami`

**Comment by oakad** [comment]
@dmitshur 
`exec.lookPath` will explicitly ignore relative paths within PATH when looking for executables, per the default value of "execerrdot" setting.

Build tools, such as Bazel, prefer to use relative paths for everything (great many reasons here, applicable to all build tools, inclusive of original "make").

Moreover, nobody guarantees that `lookPath` behavior will not change some more in the future.

To ensure stability against such changes, toolchains often implement ways to override the system PATH and rules associated with looking it up. GCC, for example, has COMPILER_PATH which takes precedence over system PATH lookup.

Go already has GOROOT; postulating that various tools always invoke go command and related binaries relative to current value of GOROOT, which itself can be relative path, may be sufficient to resolve this and many related issues.

**Comment by dmitshur** [comment]
@oakad It is a valid alternative approach to put more emphasis on GOROOT and rely on relative paths to find the go command. Migrating to that would be a large change and would need a proposal, so better to pursue outside of this issue if you'd like. I think it has already been considered fairly recently so it'd need new information to revisit. The decision made so far settled on deriving the GOROOT from the "go" command found in PATH, not the other way around. For example, see the documentation for https://pkg.go.dev/runtime#GOROOT that says "Use the system path to locate the “go” binary, and use “go env GOROOT” to find its GOROOT." per accepted proposal #51473.

**Comment by oakad** [comment]
The core issue is ability to specify the desired go toolchain in a simple and safe fashion. For example:

1. Good: using explicit path to go on the command line:
```
% mytool --go=/my/toolchain/go
```
2. Also good: using a dedicated env var, such as GOROOT:
```
% GOROOT=/my/toolchain mytool
```
3. Real bad: trusting the common PATH to do the right thing. Requiring people with complex build setups to somehow figure out what belongs on a long PATH and what does not to have the correct Go toolchain in  the mix.
